cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,623 @@
+  OTR,
+  /// Total number of symbol groups amongst which to differentiate
+  NUM_PDA_INPUT_SGS
+};
+
+/**
+ * @brief Symbols in the stack alphabet
+ */
+enum STACK_SGID : PdaStackSymbolGroupIdT {
+  /// Symbol representing the JSON-root (i.e., we're at nesting level '0')
+  STACK_ROOT = 0,
+
+  /// Symbol representing that we're currently within a list object
+  STACK_LIST = 1,
+
+  /// Symbol representing that we're currently within a struct object
+  STACK_STRUCT = 2,
+
+  /// Total number of symbols in the stack alphabet
+  NUM_STACK_SGS
+};
+
+/// Total number of symbol groups to differentiate amongst (stack alphabet * input alphabet)
+constexpr PdaSymbolGroupIdT NUM_PDA_SGIDS = NUM_PDA_INPUT_SGS * NUM_STACK_SGS;
+
+/// Mapping a input symbol to the symbol group id
+static __constant__ PdaSymbolGroupIdT tos_sg_to_pda_sgid[] = {
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, WSP, WSP, OTR, OTR, WSP, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, WSP, OTR, QTE, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, CMA, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, CLN, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OBT, ESC, CBT, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR,
+  OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OTR, OBC, OTR, CBC, OTR};
+
+/**
+ * @brief Maps a (top-of-stack symbol, input symbol)-pair to a symbol group id of the DVPA
+ */
+struct PdaSymbolToSymbolGroupId {
+  template <typename SymbolT, typename StackSymbolT>
+  __device__ __forceinline__ PdaSymbolGroupIdT
+  operator()(thrust::tuple<SymbolT, StackSymbolT> symbol_pair)
+  {
+    // The symbol read from the input
+    auto symbol = thrust::get<0>(symbol_pair);
+
+    // The stack symbol (i.e., what is on top of the stack at the time the input symbol was read)
+    // I.e., whether we're reading in something within a struct, a list, or the JSON root
+    auto stack_symbol = thrust::get<1>(symbol_pair);
+
+    // The stack symbol offset: '_' is the root group (0), '[' is the list group (1), '{' is the
+    // struct group (2)
+    int32_t stack_idx =
+      (stack_symbol == '_') ? STACK_ROOT : ((stack_symbol == '[') ? STACK_LIST : STACK_STRUCT);
+
+    // The relative symbol group id of the current input symbol
+    PdaSymbolGroupIdT symbol_gid = tos_sg_to_pda_sgid[min(
+      static_cast<int32_t>(symbol),
+      static_cast<int32_t>(sizeof(tos_sg_to_pda_sgid) / sizeof(tos_sg_to_pda_sgid[0])) - 1)];
+    return stack_idx * NUM_PDA_INPUT_SGS + symbol_gid;
+  }
+};
+
+// The states defined by the pushdown automaton
+enum pda_state_t : StateT {
+  PD_BOV,
+  PD_BOA,
+  PD_LON,
+  PD_STR,
+  PD_SCE,
+  PD_PVL,
+  PD_BFN,
+  PD_FLN,
+  PD_FNE,
+  PD_PFN,
+  PD_ERR,
+  PD_NUM_STATES
```

Comment:
Thanks for the effort of visualizing the state machine. 


In general it makes it easier to validate that input formats are in fact valid.

> Does the separation between BOV/PVL and BOA/BFN have a purpose beyond preventing trailing commas {,} or [,]?

Preventing trailing commas and accepting empty `struct` and `array` (i.e., `{}` `[]`).

> Can't we infer the difference -- whether to expect a label or not -- from the stack context?), but those questions should be easy to clear up :)

My initial thought would be that it generally should be possible. However, I'm inclined to avoid experimenting with the logic given its risk at this point. We can keep that in the back of our heads, though.